File: FindKPairswithSmallestSums.java
public class Solution {
    public List<int[]> kSmallestPairs(int[] nums1, int[] nums2, int k) {
        //currect way
        PriorityQueue<int[]> que = new PriorityQueue<>((a,b)-> a[0]+a[1]-b[0]-b[1] );
        List<int[]> res = new ArrayList<int[]>();
        if(nums1.length == 0 || nums2.length == 0 || k ==0)
            return res;
        for(int i=0;i<nums1.length && i<k;i++) que.offer(new int[]{nums1[i], nums2[0], 0});
        while(k-- >0 && !que.isEmpty() ){
            int[] arr = que.poll();
            res.add(new int[]{arr[0], arr[1]} );
            if(arr[2]<nums2.length-1){
                que.offer(new int[]{arr[0], nums2[arr[2]+1], arr[2]+1} );
            }
        }
        return res;
        
    //can't be doing in two pointers, it add two length up, not times each other.
    //need to revise it into stack way
        // List<int[]> res = new ArrayList<int[]>();
        // if(nums1.length == 0 || nums2.length == 0 )
        //     return res;
        // int l = 0, r = 0;
        // int[] pair = null;
        // while(l<nums1.length && r<nums2.length && l+r < k){
        //     pair = new int[2];
        //     pair[0] = nums1[l];
        //     pair[1] = nums2[r];
        //     res.add(pair);
        //     if(l+1<nums1.length && r+1<nums2.length){
        //         if(nums1[l+1] - nums1[l] > nums2[r+1]- nums2[r]){
        //             r++;
        //         }
        //         else{
        //             l++;
        //         }
        //     }
        //     else if(l+1==nums1.length){
        //         r++;
        //     }
        //     else if(r+1== nums2.length){
        //         l++;
        //     }
        // }
        // return res;
    }
}

File: findKPairsWithSmallestSums.kt
// regarding to tutorial, keep an hashset to dedupe pair of indexes,
// starting with 0,0 pop it from pq and add two more possible least big pairs as 1,0 and 0,1 
// then pop another one and add +1,0 and 0,+1 to the pop out pair, keep the cycle going until find k or exhaust two arrays
class Solution {
    fun kSmallestPairs(nums1: IntArray, nums2: IntArray, k: Int): List<List<Int>> {
        val pq = PriorityQueue<Pair<Int, Int>>(compareBy {nums1[it.first] + nums2[it.second]})
        val hset = HashSet<Pair<Int, Int>>()
        val res = mutableListOf<List<Int>>()
        
        val m = nums1.size
        val n = nums2.size
        pq.add(Pair(0, 0))
        hset.add(Pair(0, 0))
        while(res.size < k && pq.isNotEmpty()){
            val (idx1, idx2) = pq.poll()
            res.add(listOf(nums1[idx1], nums2[idx2]))
            for((x, y) in setOf(1 to 0, 0 to 1)){
                val n1 = idx1 + x
                val n2 = idx2 + y
                if(n1 < m && n2 < n && !hset.contains(Pair(n1, n2)) ){
                    pq.add(Pair(n1, n2))
                    hset.add(Pair(n1, n2))
                }
            }
        }
        return res
    }
}
